src/server/pfs/server/driver.go

```diff
@@ -3310,28 +3285,76 @@ func (s *putFileServer) Peek() (*pfs.PutFileRequest, error) {
 	return req, nil
 }
 
-func forEachPutFile(server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) error {
+func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) (bool, string, string, error) {
 	limiter := limit.New(client.DefaultMaxConcurrentStreams)
 	var pr *io.PipeReader
 	var pw *io.PipeWriter
 	var req *pfs.PutFileRequest
 	var err error
 	var eg errgroup.Group
+	oneOff := false
+	var rawCommitID string
+	var commitID string
+	var repo string
+	var branch string
+
 	for req, err = server.Recv(); err == nil; req, err = server.Recv() {
 		req := req
 		if req.File != nil {
+			// For the first file, dereference the commit ID if needed. For
+			// subsequent files, ensure that they're all referencing the same
+			// commit ID, and replace their commit objects to all refer to the
+			// same thing.
+			if rawCommitID == "" {
+				commit := req.File.Commit
+				repo = commit.Repo.Name
+				// The non-dereferenced commit ID. Used to ensure that all
+				// subsequent requests use the same value.
+				rawCommitID = commit.ID
+				// oneOff is true if we're creating the commit as part of this put-file
+				oneOff = false
+				// inspectCommit will replace file.Commit.ID with an actual commit ID if
+				// it's a branch. So we want to save it first.
+				if !uuid.IsUUIDWithoutDashes(commit.ID) {
+					branch = commit.ID
+				}
+				commitInfo, err := d.inspectCommit(pachClient, commit, pfs.CommitState_STARTED)
+				if err != nil {
+					if (!isNotFoundErr(err) && !isNoHeadErr(err)) || branch == "" {
+						return oneOff, repo, branch, err
+					}
+					oneOff = true
+				}
+				if commitInfo != nil && commitInfo.Finished != nil {
+					if branch == "" {
+						err = pfsserver.ErrCommitFinished{commit}
+						return oneOff, repo, branch, err
+					}
+					oneOff = true
+				}
+				commitID = commit.ID
+			} else if req.File.Commit.ID != rawCommitID {
+				err = fmt.Errorf("All requests in a put files call must have the same commit ID; expected '%s', got '%s'", rawCommitID, req.File.Commit.ID)
+				return oneOff, repo, branch, err
```

Comment:
I'm not sure if this would break `putFiles`, but would it make sense to return `false, "", "", err` here (and in error cases below)? My impression is that it's more conventional in go to discard other return values when returning an error.